### ISSUE TITLE
Inventory all of the TODOs

### ISSUE THREAD

**Issue by estherkim** [ampproject/amphtml#19767]

**Comment by rsimha** [comment]
Here are the `TODO`s in `build-system/`:

```
app-index/components/main.js:17:// TODO(alanorozco): This bundle shall be used for experiment toggles.
eslint-rules/todo-format.js:23:          if (/TODO/.test(comment.value) &&
eslint-rules/todo-format.js:24:                !/TODO\(@?\w+,\s*#\d{1,}\)/.test(comment.value)) {
eslint-rules/todo-format.js:27:              message: 'TODOs must be in TODO(@username, #1234) format. ' +
tasks/lint.js:81:    // TODO(#15255, #14761): Remove log folding after warnings are fixed.
tasks/lint.js:104:        // TODO(#15255, #14761): Remove log folding after warnings are fixed.
tasks/create-golden-css/css/main.css:248:  /* TODO(dvoytenko): explore adding here object-fit. */
tasks/create-golden-css/css/main.css:386: * TODO(dvoytenko, #8464): cleanup old `i-amp-scroll-container`.
tasks/create-golden-css/css/main.css:400: * TODO(dvoytenko, #8464): cleanup old `i-amp-scroll-container`.
tasks/presubmit-checks.js:48:  // TODO(dvoytenko, #8464): cleanup whitelist.
tasks/changelog.js:295:      // TODO: figure out how to break down validator changes since
tasks/csvify-size/index.js:154:    // TODO(erwinm): figure out where this is occurring.
tasks/dep-check.js:185:  // TODO(erwinm): Try and work this in with `gulp build` so that
tasks/runtime-test.js:104:        // TODO(amp-infra): Restore this once tests are stable again.
tasks/runtime-test.js:108:        // TODO(amp-infra): Evaluate and add more platforms here.
tasks/runtime-test.js:118:      // TODO(rsimha): Add more browsers to this list. #6039.
tasks/runtime-test.js:571:  // TODO(rsimha, 14814): Remove after Karma / Sauce ticket is resolved.
tasks/runtime-test.js:736:  // TODO(alanorozco): Come up with a more elegant check?
tasks/compile.js:147:    // TODO(@cramforce): Run the post processing step
tasks/bundle-size.js:368:    // TODO(danielrozenberg): remove the legacy check once the app has been
tasks/karma.conf.js:56:  // TODO(rsimha, #15510): Sauce labs on Safari doesn't reliably support
tasks/todos.js:33: * Test if a file's contents contains closed TODOs.
tasks/todos.js:36: * @return {Promise<number>} Number of found closed TODOs.
tasks/todos.js:40:  const todos = contents.match(/TODO\([^\)]*\)/g);
tasks/todos.js:48:    const parts = /TODO\([^\)]*\#(\d*)\)/.exec(todo);
tasks/todos.js:134:          log(colors.red('Found closed TODOs: ', foundCount));
tasks/todos.js:141:gulp.task('todos:find-closed', 'Find closed TODOs', findClosedTodosTask);
app-video-testbench.js:113:  // TODO(alanorozco): Reenable with valid params, if possible.
app-video-testbench.js:299:  // TODO(alanorozco): Use config.urls.cdn value. This file is not available
single-pass.js:51:// TODO(rsimha, erwinmombay): Derive this list programmatically if possible.
single-pass.js:160:  // TODO(erwinm): special case src/amp.js for now. add a propert sort
single-pass.js:170:    // TODO(erwinm): This access will break
single-pass.js:188:      // TODO(@cramforce): Remove special case.
single-pass.js:538: * TODO(#18811, erwinm): this breaks source maps and we need a way to fix this.
single-pass.js:559: * TODO(erwinm, #18811): This operation is needed but straight out breaks
typescript.js:46:  // TODO(choumx): This was partially copy-pasta'd from tsickle. Add a default
pr-check.js:407:        // TODO(choumx, #19658): --headless disabled for integration tests on
pr-check.js:422:    // TODO(choumx, #19658): --headless disabled for integration tests on
app-utils.js:27:    // TODO:(ccordry) remove this when story 0.1 is deprecated
app.js:43:// TODO: shall we turn on CSP all the time?
app.js:152:    // TODO(dvoytenko): would be nicer to do forward instead of redirect.
app.js:404:    // TODO(ccordry): Remove this when story v01 is depricated.
amp4test.js:53:    // TODO: Why is setting localDev necessary?
amp4test.js:63:  // TODO: Do we need to inject amp-3p-iframe-src for non-ad tests?
amp4test.js:308:          // TODO: Should this be a local URL i.e. /dist/v0/...?
amp.extern.js:161:// TODO(bradfrizzell) Move to its own extern. Not relevant to all AMP.
amp.extern.js:328:// TODO (remove after we update closure compiler externs)
```

**Comment by rsimha** [comment]
Let's try knocking out the ones that remain during fixit week.

```
rsimha@bobdylan (master):~/src/amphtml/build-system$ cs TODO
app-index/amphtml-helpers.js:32:// TODO(alanorozco): Expand
app-index/amphtml-helpers.js:120:  // TODO(alanorozco): Too greedy. Parse "on" attributes instead.
app-index/components/main.js:17:// TODO(alanorozco): This bundle shall be used for experiment toggles.
eslint-rules/todo-format.js:24:            /TODO/.test(comment.value) &&
eslint-rules/todo-format.js:25:            !/TODO\(@?\w+,\s*#\d{1,}\)/.test(comment.value)
eslint-rules/todo-format.js:30:                'TODOs must be in TODO(@username, #1234) format. ' +
tasks/extension-helpers.js:470:      // TODO(choumx): Compile this when worker-dom externs are available.
tasks/runtime-test/helpers.js:41:  //TODO(estherkim): consolidate arg validation logic
tasks/runtime-test/index.js:125:          // TODO(amp-infra): Evaluate and add more platforms here.
tasks/runtime-test/index.js:447:  // TODO(rsimha, 14814): Remove after Karma / Sauce ticket is resolved.
tasks/runtime-test/index.js:665:  // TODO(alanorozco): Come up with a more elegant check?
tasks/create-golden-css/css/main.css:248:  /* TODO(dvoytenko): explore adding here object-fit. */
tasks/create-golden-css/css/main.css:386: * TODO(dvoytenko, #8464): cleanup old `i-amp-scroll-container`.
tasks/create-golden-css/css/main.css:400: * TODO(dvoytenko, #8464): cleanup old `i-amp-scroll-container`.
tasks/presubmit-checks.js:51:  // TODO(dvoytenko, #8464): cleanup whitelist.
tasks/presubmit-checks.js:312:      // TODO(#22414): Remove paths as they are migrated off of sync API.
tasks/changelog.js:314:      // TODO: figure out how to break down validator changes since
tasks/csvify-size/index.js:165:      // TODO(erwinm): figure out where this is occurring.
tasks/e2e/amp-driver.js:44:      // TODO(estherkim): somehow allow non-8000 port and domain
tasks/e2e/amp-driver.js:54:      // TODO(cvializ): this is a HACK
tasks/e2e/amp-driver.js:63:      // TODO(estherkim): somehow allow non-8000 port and domain
tasks/e2e/puppeteer-controller.js:537:   * TODO(cvializ): decide if we need to waitForNavigation on click and keypress
tasks/e2e/describes-e2e.js:122:  // TODO(estherkim): implement sessions
tasks/e2e/describes-e2e.js:150:  // TODO(cvializ,estherkim,sparhami):
tasks/e2e/selenium-webdriver-controller.js:481:    // TODO(sparhami) These are throwing errors, but are not causing the test
tasks/e2e/selenium-webdriver-controller.js:603:    // TODO(cvializ): Errors? Or maybe ChromeDriver hasn't yet implemented
tasks/e2e/selenium-webdriver-controller.js:615:    // TODO(estherkim): add 'id' parameter, to select element inside 'handle'
tasks/e2e/selenium-webdriver-controller.js:619:    // TODO(cvializ): This trashes cached element handles why?
tasks/dep-check.js:206:  // TODO(erwinm): Try and work this in with `gulp build` so that
tasks/karma.conf.js:61:  // TODO(rsimha, #15510): Sauce labs on Safari doesn't reliably support
tasks/karma.conf.js:293:  // TODO(rsimha): Reduce this number once keepalives are implemented by
tasks/todos.js:32: * Test if a file's contents contains closed TODOs.
tasks/todos.js:35: * @return {Promise<number>} Number of found closed TODOs.
tasks/todos.js:39:  const todos = contents.match(/TODO\([^\)]*\)/g);
tasks/todos.js:47:    const parts = /TODO\([^\)]*\#(\d*)\)/.exec(todo);
tasks/todos.js:136:        log(colors.red('Found closed TODOs: ', foundCount));
tasks/todos.js:146:todosFindClosed.description = 'Find closed TODOs';
app-video-testbench.js:113:  // TODO(alanorozco): Reenable with valid params, if possible.
app-video-testbench.js:299:  // TODO(alanorozco): Use config.urls.cdn value. This file is not available
typescript.js:52:  // TODO(choumx): This was partially copy-pasta'd from tsickle. Add a default
dep-check-config.js:260:      // TODO(ccordry): remove this after createShadowRootWithStyle is moved to src
app-utils.js:27:    // TODO:(ccordry) remove this when story 0.1 is deprecated
app.js:55:// TODO: shall we turn on CSP all the time?
app.js:221:    // TODO(dvoytenko): would be nicer to do forward instead of redirect.
app.js:543:    // TODO(ccordry): Remove this when story v01 is depricated.
amp4test.js:62:    // TODO: Why is setting localDev necessary?
amp4test.js:71:  // TODO: Do we need to inject amp-3p-iframe-src for non-ad tests?
amp4test.js:331:          // TODO: Should this be a local URL i.e. /dist/v0/...?
compile/single-pass.js:68:// TODO(rsimha, erwinmombay): Derive this list programmatically if possible.
compile/single-pass.js:182:  // TODO(erwinm): special case src/amp.js for now. add a propert sort
compile/single-pass.js:192:    // TODO(erwinm): This access will break
compile/single-pass.js:210:      // TODO(@cramforce): Remove special case.
compile/single-pass.js:592: * TODO(#18811, erwinm): this breaks source maps and we need a way to fix this.
compile/single-pass.js:616: * TODO(erwinm, #18811): This operation is needed but straight out breaks
compile/single-pass.js:656:  // TODO(@cramforce): Run the post processing step
compile/closure-compile.js:81:    // TODO(rsimha): Use the native compiler instead of compiler.jar once a fix
amp.extern.js:176:// TODO(bradfrizzell) Move to its own extern. Not relevant to all AMP.
amp.extern.js:369:// TODO (remove after we update closure compiler externs)
```

**Comment by rsimha** [comment]
Hah! The list has grown.

_Wistfully remembering the days when it was easy to fly to MTV for fix-it week_

```
rsimha@rsimha-macbookpro2 (master):~/src/amphtml/build-system$ cs TODO .
./tasks/server-tests.js:80:  // TODO(rsimha): Change require to import when node v14 is the active LTS.
./tasks/csvify-size/index.js:166:      // TODO(erwinm): figure out where this is occurring.
./tasks/build.js:41: * TODO(rsimha, #27471): Remove this after several weeks.
./tasks/check-owners.js:65:  // TODO(rcebulko): Replace this placeholder with a check that uses the
./tasks/todos.js:33: * Test if a file's contents contains closed TODOs.
./tasks/todos.js:36: * @return {Promise<number>} Number of found closed TODOs.
./tasks/todos.js:40:  const todos = contents.match(/TODO\([^\)]*\)/g);
./tasks/todos.js:48:    const parts = /TODO\([^\)]*\#(\d*)\)/.exec(todo);
./tasks/todos.js:138:        log(colors.red('Found closed TODOs: ', foundCount));
./tasks/todos.js:148:todosFindClosed.description = 'Find closed TODOs';
./tasks/karma.conf.js:59:  // TODO(rsimha, #15510): Sauce labs on Safari doesn't reliably support
./tasks/presubmit-checks.js:51:  // TODO(dvoytenko, #8464): cleanup whitelist.
./tasks/presubmit-checks.js:327:      // TODO(#22414): Remove paths as they are migrated off of sync API.
./tasks/presubmit-checks.js:756:      // Test files. TODO(#24144): Fix these and remove from the whitelist.
./tasks/runtime-test/runtime-test-base.js:62:          // TODO(amp-infra): Evaluate and add more platforms here.
./tasks/helpers.js:380:  // TODO: @jonathantyng remove browserifyOptions #22757
./tasks/performance/rewrite-analytics-tags.js:51:    // TODO (micajuineho) replace with analytics/config.js merge objects
./tasks/e2e/describes-e2e.js:45: * TODO(cvializ): Firefox now experimentally supports puppeteer.
./tasks/e2e/describes-e2e.js:180:      //TODO(estherkim): workaround. `onQuit()` was added in selenium-webdriver v4.0.0-alpha.5
./tasks/e2e/puppeteer-controller.js:575:   * TODO(cvializ): decide if we need to waitForNavigation on click and keypress
./tasks/e2e/amp-driver.js:62:      // TODO(estherkim): somehow allow non-8000 port and domain
./tasks/e2e/amp-driver.js:72:      // TODO(cvializ): this is a HACK
./tasks/e2e/amp-driver.js:84:      // TODO(estherkim): somehow allow non-8000 port and domain
./tasks/e2e/selenium-webdriver-controller.js:523:    // TODO(sparhami) These are throwing errors, but are not causing the test
./tasks/e2e/selenium-webdriver-controller.js:626:    // TODO(cvializ): Errors? Or maybe ChromeDriver hasn't yet implemented
./tasks/e2e/selenium-webdriver-controller.js:647:    // TODO(estherkim): add 'id' parameter, to select element inside 'handle'
./tasks/e2e/selenium-webdriver-controller.js:651:    // TODO(cvializ): This trashes cached element handles why?
./tasks/create-golden-css/css/main.css:248:  /* TODO(dvoytenko): explore adding here object-fit. */
./tasks/create-golden-css/css/main.css:386: * TODO(dvoytenko, #8464): cleanup old `i-amp-scroll-container`.
./tasks/create-golden-css/css/main.css:400: * TODO(dvoytenko, #8464): cleanup old `i-amp-scroll-container`.
./tasks/dep-check.js:210:  // TODO(erwinm): Try and work this in with `gulp build` so that
./externs/amp.extern.js:183:// TODO(bradfrizzell) Move to its own extern. Not relevant to all AMP.
./externs/amp.extern.js:392:// TODO (remove after we update closure compiler externs)
./externs/amp.extern.js:860: * TODO(dvoytenko): remove FeaturePolicy once it's added in Closure externs.
./server/amp4test.js:54:    // TODO: Why is setting localDev necessary?
./server/amp4test.js:63:  // TODO: Do we need to inject amp-3p-iframe-src for non-ad tests?
./server/amp4test.js:332:          // TODO: Should this be a local URL i.e. /dist/v0/...?
./server/app-index/components/main.js:17:// TODO(alanorozco): This bundle shall be used for experiment toggles.
./server/app-index/amphtml-helpers.js:32:// TODO(alanorozco): Expand
./server/app-index/amphtml-helpers.js:120:  // TODO(alanorozco): Too greedy. Parse "on" attributes instead.
./server/app-video-testbench.js:114:  // TODO(alanorozco): Reenable with valid params, if possible.
./server/app-video-testbench.js:300:  // TODO(alanorozco): Use config.urls.cdn value. This file is not available
./server/app-utils.js:96:    // TODO:(ccordry) remove this when story 0.1 is deprecated
./server/app.js:57:// TODO(#24333): Migrate all server URL handlers to new-server/router and
./server/app.js:70:// TODO: shall we turn on CSP all the time?
./server/app.js:239:    // TODO(dvoytenko): would be nicer to do forward instead of redirect.
./server/app.js:506:    // TODO(ccordry): Remove this when story v01 is depricated.
./babel-config/pre-closure-config.js:64:    // TODO(alanorozco): Remove `replaceCallArguments` once serving infra is up.
./common/check-package-manager.js:314:  // Python 3.5+ are still supported (TODO: deprecate 3.5 on 2020-09-13)
./pr-check/module-dist-bundle-size.js:51:    // TODO(#27703): Run bundle size check (--on_push_build)
./pr-check/module-dist-bundle-size.js:59:      // TODO(#27703): Run bundle size check (--on_pr_build)
./compile/typescript.js:52:  // TODO(choumx): This was partially copy-pasta'd from tsickle. Add a default
./compile/single-pass.js:174:  // TODO(erwinm): special case src/amp.js for now. add a propert sort
./compile/single-pass.js:184:    // TODO(erwinm): This access will break
./compile/single-pass.js:202:      // TODO(@cramforce): Remove special case.
./compile/single-pass.js:548: * TODO This should operate on the gulp stream, not on disk files.
./compile/single-pass.js:599: * TODO This should operate on the gulp stream, not on disk files.
./compile/single-pass.js:620: * TODO This should operate on the gulp stream, not on disk files.
./compile/single-pass.js:695:  // TODO(@cramforce): Run the post processing step
./compile/single-pass.js:719: * TODO This should operate on the gulp stream, not on disk files.
./compile/sources.js:94:  // TODO(#24080) Remove this when story ads have full ad network support.
./babel-plugins/babel-plugin-transform-amp-asserts/index.js:74:      // TODO(jridgewell): enable this later.
./babel-plugins/babel-plugin-transform-inline-configure-component/index.js:209:        // TODO(go.amp.dev/issue/26948): sourcemaps
./eslint-rules/no-arrow-on-register-functions.js:28:          // TODO(erwinm): add fixer method.
./eslint-rules/todo-format.js:24:            /TODO/.test(comment.value) &&
./eslint-rules/todo-format.js:25:            !/TODO\(@?\w+,\s*#\d{1,}\)/.test(comment.value)
./eslint-rules/todo-format.js:30:                'TODOs must be in TODO(@username, #1234) format. ' +
./eslint-rules/window-property-name.js:29:// TODO(choumx): Fix extensions code and add 'extensions/' here.
./test-configs/dep-check-config.js:246:      // TODO(#24080) Remove this when story ads have full ad network support.
./test-configs/dep-check-config.js:248:      // TODO(ccordry): remove this after createShadowRootWithStyle is moved to src
rsimha@rsimha-macbookpro2 (master):~/src/amphtml/build-system$ 
```

**Comment by samouri** [comment]
is this still relevant?

**Comment by estherkim** [comment]
Yep, pretty sure we haven't eliminated all of them.

**Comment by rsimha** [comment]
_[Infra triage]_ I scanned through several of the TODOs in `build-system/` and noticed that the reasons behind adding most of them haven't entirely gone away. It's probably okay for them to remain and serve as documentation that will help during future rewrites / refactors of code.

While I acknowledge the original intent of this issue (to inventory and fix as many TODOs as possible), I think it's unlikely that we'll ever solve this problem via a single tracking issue. Closing this in favor of organically implementing targeted fixes.
